### PR TITLE
feat: enable IAM permissions boundaries

### PR DESCRIPTION
**Pull request by huanjani**
Related: #2986.

This is the first PR in the IAM permissions boundary feature.

Some companies leverage AWS Organizations's [Service Control Policies](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_scps.html) (SCP) feature, which can be used together with [IAM Permissions Boundaries](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html) (see diagram below) to [limit](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_scps.html#scp-effects-on-permissions) the actions devs may perform and the resources they may affect. 
image

By allowing users to indicate their existing permissions boundary policies when running `app init`, Copilot can add them to IAM Roles when creating them. This PR conditionally adds a boundary policy to the IAM roles created at the app level. More IAM Role modifications to come.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/template/templates/app/app.yml [hunk 1]
```diff
@@ -42,6 +42,9 @@ Resources:
               Service: cloudformation.amazonaws.com
             Action:
               - sts:AssumeRole
+{{- if .PermissionsBoundary }}
+      PermissionsBoundary: !Sub 'arn:${AWS::Partition}:iam::${AWS::AccountId}:policy/{{.PermissionsBoundary}}'
```
Comment (Lou1415926) [inline internal/pkg/template/templates/app/app.yml:46]: Just to make sure - we are assuming that the PermissionBoundary lives in the account that deploys the app, is that right?

**Review by Lou1415926 — APPROVED** [review]
LGTM! We probably will need to update some integration test too 🤔

File: internal/pkg/template/templates/app/app.yml (hunk 1 above)
Comment (huanjani) [reply internal/pkg/template/templates/app/app.yml:46]: For these IAM roles, yes. If they deploy cross-account environments, we'll refer to those account IDs.

Comment (dannyrandall) [reply internal/pkg/template/templates/app/app.yml:46]: does that mean they'd need a copy of the permission boundary in every account they deploy in to? would it make sense to let them specify the full ARN and reference a permission boundary in a specific account? (i have no idea if that's possible, just checking)

**Review by dannyrandall — APPROVED** [review]
looks good, just a question!

File: internal/pkg/template/templates/app/app.yml (hunk 1 above)
Comment (efekarakus) [reply internal/pkg/template/templates/app/app.yml:46]: They'd duplicate the same managed policy in all the accounts! The policy name should be the same in the app and env accounts as well (that's why we can safely use the flag value for _all_ our iam roles)

Comment (iamhopaul123) [reply internal/pkg/template/templates/app/app.yml:46]: We might need to upgrade the app template version as well 👀

Comment (huanjani) [reply internal/pkg/template/templates/app/app.yml:46]: upgraded app template version. thanks!

**Pull request merged**